li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/producer/LiKafkaFederatedProducerImpl.java

```diff
@@ -355,7 +361,68 @@ public void close() {
 
   @Override
   public void close(long timeout, TimeUnit timeUnit) {
-    throw new UnsupportedOperationException("Not implemented yet");
+    if (_producers.isEmpty()) {
+      LOG.warn("No producers to close for cluster group {}", _clusterGroup);
+      return;
+    }
+
+    LOG.info("Closing LiKafkaProducer for cluster group {} in {} {}...", _clusterGroup, timeout, timeUnit);
+
+    long startTimeMs = System.currentTimeMillis();
+    long budgetMs = timeUnit.toMillis(timeout);
+    long deadlineTimeMs = startTimeMs + budgetMs;
+
+    List<Thread> taskThreads = new ArrayList<>();
+    for (Map.Entry<ClusterDescriptor, PerClusterProducerContext<K, V>> entry : _producers.entrySet()) {
+      Thread taskThread = new Thread(new CloseTask(entry.getValue(), deadlineTimeMs));
+      taskThreads.add(taskThread);
+      taskThread.start();
+    }
+
+    for (Thread taskThread : taskThreads) {
+      try {
+        taskThread.join();
+      } catch (InterruptedException e) {
+        throw new KafkaException("Fail to close all producers for cluster group " + _clusterGroup, e);
+      }
+    }
+
+    LOG.info("LiKafkaProducer close for cluster group {} complete in {} milliseconds", _clusterGroup,
+        (System.currentTimeMillis() - startTimeMs));
+  }
+
+  private class CloseTask implements Runnable {
+    private PerClusterProducerContext<K, V> _context;
+    private long _deadlineTimeMs;
+
+    CloseTask(PerClusterProducerContext<K, V> context, long deadlineTimeMs) {
+      _context = context;
+      _deadlineTimeMs = _deadlineTimeMs;
+    }
+
+    public void run() {
+      LOG.info("Closing LiKafkaProducer for cluster {}", _context.cluster());
+      LongAdder numThreadsInSend = _context.numThreadsInSend();
+
+      _context.setClosed(true);
+      synchronized (numThreadsInSend) {
```

Comment:
I am not sure why we need synchronize here if `numThreadsInSend` is per-producer context?

Reply:
I think we should put notifyAll() in a synchronized block, otherwise a notification may be lost due to a race condition.